### ISSUE TITLE
Custom HTML assembly (user completely owning HTML assembly, instead of `render`/`onRenderHtml`/`escapeInject`/`dangerouslySkipEscape`)

### ISSUE THREAD

**Issue by brillout** [brillout/vite-plugin-ssr#545]
### Description

Some users want/need more control over the HTML than currently possible.

If you find yourself in that situation, write a comment here and explain why you want this.

**Comment by coryvirok** [comment]
+1 on this. I would like to be able to run the HTML through a minifier in `_default.page.server.js#render()`

**Comment by brillout** [comment]
Alright, I had a deep look at this.

I tried, I really did, but I've come to the conclusion that it's best to not support this.

I do understand that, as a VPS user, it's very tempting to want to use VPS merely as a "page router" while manually stiching the HTML myself. It sounds very sexy at first, but I've come to the conclusion that it's the wrong approach.

The thing is that properly stiching the various HTML parts together is a *lot* more complex than one might think. There are a *lot* of edge cases. The HTML parts are assembled fundamentally differently in dev than in production (in dev the bottleneck is transpilation, whereas in prod the bottleneck is network latency). Another example is HTML streaming which is a whole topic on its own that is non-trivial to say the least (e.g. depending on various situations, the JSON data needs to be injected before or after the stream ends, or even sometimes during the stream).

Enabling custom HTML assembly is opening a can of worms and AFAICT we'll collectively end up regreting this, including users who end up with a half-assed HTML assembly with major performance degradations.

It's not a definite no: if someone comes up with a *concrete* and valid reason then I'm open for changing my mind.

Until then, it's best to let VPS *completely* own HTML assembly.

@gajus I looked at your code and I do believe it's best for Contra to let VPS assemble the HTML.

```ts
// startHttpService.ts

import { renderPage as renderVitePage } from 'vite-plugin-ssr';

app.get('*', async (request, reply) => {
   const htmlContent = {
     head: getHtmlHead(request),
     body: getHtmlBody(request)
   };

   const pageContextInit = {
     // I know about your `viteUrl` hack but I believe the new VPS configs
     // baseAssets and baseServer shoud fix your poblem
     urlOriginal: request.url,
     // Pass your custom html tags here
     htmlContent
   };
   const pageContext = await renderVitePage(pageContextInit)

   if (!pageContext.httpResponse) {
     /* ... */
   }

   let html = pageContext.httpResponse.body;
   // Instead of generateHtml()
   html = inlineJS(html);
   html = '<!DOCTYPE html>' + html;

   const { statusCode, contentType } = pageContext.httpResponse;
   void reply
    .code(statusCode)
    .type('text/html')
    .send(html);
});

function inlineJS(html: string): string {
  // Parse the HTML and replace <script src="/script.js"> with <script>${scriptContent}</script>
  // (This optimization could actually be done by default by VPS for SPA pages.)
}

function getHtmlHead(request): string {
   let head = '';

   const getGoogleFontCss = getGoogleFontCss();
   head += `<style>${googleFontsCss}</style>`;

   const urlMetaData = getUrlMetaData(request);
   head += `
     <title>${escapeHtml(urlMetaData.title)}</title>
     <meta content="${urlMetaData.canonicalUrl}" property="og:url" />
     <link rel="canonical" href="${urlMetaData.canonicalUrl}" />
     <meta content="${escapeHtml(urlMetaData.description)}" name="description" />
     <meta content="${escapeHtml(
       urlMetaData.description
     )}" name="twitter:description" />
     <meta content="${escapeHtml(
       urlMetaData.description
     )}" property="og:description" />
     <meta content="${escapeHtml(urlMetaData.title)}" name="twitter:title" />
     <meta content="${escapeHtml(urlMetaData.title)}" property="og:title" />
     <meta content="${urlMetaData.facebookImageUrl}" property="og:image"/>
     <meta content="${urlMetaData.twitterImageUrl}" name="twitter:image"/>
  `;

  return head;
}
function getHtmlBody(request): string {
   let body = '';
   body += '<div id="react-container" />';

   const nextAppConfiguration = getNextAppConfiguration();
   body += `
     <script>
       var global = global || window;
       window.CWA = ${stringify(nextAppConfiguration)};
     </script>
   `;

   const appConfiguration = getAppConfiguration();
   if (appConfiguration.googleMeasurementId) {
     const { googleMeasurementId } = appConfiguration;
     body += `
       <script async src="https://www.googletagmanager.com/gtag/js?id=${googleMeasurementId}"></script>
       <script>
         window.dataLayer = window.dataLayer || [];
         function gtag(){dataLayer.push(arguments);}
         gtag('js', new Date());
         gtag('config', '${googleMeasurementId}');
       </script>
     `;
   }

   return body;
}
```

```ts
// renderer/_default.page.server.ts

export { render };

import { escapeInject, dangerouslySkipEscape } from 'vite-plugin-ssr';

function render(pageContext: { htmlContent: { head: string, body: string }}) {
  return escapeInject`
    <html>
      <head>
        ${dangerouslySkipEscape(pageContext.htmlContent.head)}
      </head>
      <body>
        ${dangerouslySkipEscape(pageContext.htmlContent.body)}
      </body>
    </html>
  `;
}
```

I don't know much about your code, but AFAICT this should work.

This appraoch may lead to some (minor) clashes with how VPS injects its assets, but I've ideas how to fix these potential problems. I'd suggest we try and see how it goes.

I understand it's ugly to have `/renderer/_default.page.server.js` be the only server-side code that is transpiled by Vite and the only reason for having a `dist/server/`, but, all-in-all, I believe this ugliness to be very well worth it.

**Comment by gajus** [comment]
Just coming back to this as I am working on a related refactor.

We ended up not using this, as our setup is just too far from how `vite-plugin-ssr` expects it to be setup.

I am quite happy with our current setup for what it is worth.

The only part that would be nice to get rid of is this though:

```ts
/**
 * At the moment, vite-plugin-ssr does not provide a way to get `pageContext`,
 * in the context of `renderPage`. We therefore have to extract that from the generated HTML.
 */
const getClientPageContext = (body: string): string => {
  const startHook =
    '<script id="vite-plugin-ssr_pageContext" type="application/json">';

  const startIndex = body.indexOf(startHook);

  if (startIndex === -1) {
    throw new Error('pageContext start not found');
  }

  const tail = body.slice(startIndex + startHook.length);

  const endIndex = tail.indexOf('</script>');

  if (endIndex === -1) {
    throw new Error('pageContext end not found');
  }

  return tail.slice(0, endIndex);
};
```

Ideally this was returned by `renderPage`.

**Comment by brillout** [comment]
Done: you can now access `pageContext._pageContextHtmlTag` (released in `0.4.64`).

I've added [Contra integration tests](https://github.com/brillout/vite-plugin-ssr/blob/53eb119029b1c213ed4fb45b9cccbf18825b01f6/test/contra/test.ts#L10-L13) (this is only a fraction of it) to VPS's test suite, which means you'll be able to use `pageContext._pageContextHtmlTag` for the foreseeable future. The `_pageContextHtmlTag` is marked private and isn't documented, since VPS's official stance is to discourages this practice.

@gajus I understand that you're sticking to what works, but let me know if want to give this another shot and I'll happily assist (feel free to PM on Discord). Eventually VPS will do things in a more performant way than Contra. (SPA performance wasn't the number 1 focus so far, but this will change.) No rush: your current approach is CI'd.

Closing this. I'll re-open if someone comes up with a convincing reason for supporting this.
